### PR TITLE
vhost_user_backend: Make sure multiqueue/multithread can be supported without contention

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request updates the vhost_user_backend crate (and each backend implementation accordingly) so that multiqueues on multiple threads can be supported without hitting a contention point that causes no performance benefit.

It also includes the modification of the vhost-user-net backend so that multithreaded multiqueue can be supported.

### CODE REVIEW

**Comment by sboeuf** [comment]
Please don't merge this PR yet. The goal is to validate this refactoring does not cause any side effect with the CI.

**Comment by rbradford** [comment]
You probably mean contention not congestion :-)

**Comment by sboeuf** [comment]
@rbradford @sameo @bjzhjing 
Ok so the PR is ready, achieving the expected performance improvement.
I have been able to remove the contention point from the `vhost_user_backend` crate as a first step, and then I have done all the necessary changes to the `vhost-user-net` backend to support multithreaded multiqueue.
Per my testing, I got the following numbers:
```
1 vCPU, 1 iperf3 connection, 1 tx/rx pair (2 queues) => 16 Gbits/sec
2 vCPUs, 2 iperf3 connections, 2 tx/rx pairs (4 queues) => 24 Gbits/sec
4 vCPUs, 4 iperf3 connections, 4 tx/rx pairs (8 queues) => 28 Gbits/sec
```
I think we can further improve the backend crate code (not the performances though), but I'll do that in some follow up PRs only after this PR will be merged.

@bjzhjing could you give this PR a try to confirm that you can get the correct numbers with your setup too? That'd be helpful to validate I didn't do something wrong when running my tests :)

**Comment by bjzhjing** [comment]
> @rbradford @sameo @bjzhjing
> Ok so the PR is ready, achieving the expected performance improvement.
> I have been able to remove the contention point from the `vhost_user_backend` crate as a first step, and then I have done all the necessary changes to the `vhost-user-net` backend to support multithreaded multiqueue.
> Per my testing, I got the following numbers:
> 
> ```
> 1 vCPU, 1 iperf3 connection, 1 tx/rx pair (2 queues) => 16 Gbits/sec
> 2 vCPUs, 2 iperf3 connections, 2 tx/rx pairs (4 queues) => 24 Gbits/sec
> 4 vCPUs, 4 iperf3 connections, 4 tx/rx pairs (8 queues) => 28 Gbits/sec
> ```
> 
> I think we can further improve the backend crate code (not the performances though), but I'll do that in some follow up PRs only after this PR will be merged.
> 
> @bjzhjing could you give this PR a try to confirm that you can get the correct numbers with your setup too? That'd be helpful to validate I didn't do something wrong when running my tests :)

I've tried it, and gain the same performance data as I listed before, that's very nice, thanks! Only one thing I want to mention that the current factorization indeed makes the backend implementation very simple, however, vhost_user_backend itself becomes bigger and complex, it may make future maintenance a bit difficult, and loose flexibility if we'd like to do some backends specified change. Anyway, I do not see such issue currently, just a feeling :-) I will close the PR #887 and #785.

**Comment by sboeuf** [comment]
@bjzhjing 
>I've tried it, and gain the same performance data as I listed before, that's very nice, thanks!

Thank you for testing this Cathy, I feel more confident in merging this now that you validated it works for you as well.

>Only one thing I want to mention that the current factorization indeed makes the backend implementation very simple, however, vhost_user_backend itself becomes bigger and complex, it may make future maintenance a bit difficult, and loose flexibility if we'd like to do some backends specified change. Anyway, I do not see such issue currently, just a feeling :-)

Well that's okay and that's the point of this crate. As long as we can manage to factorize as much as we can into this crate, this makes the maintenance for each backend implementation easier.

>I will close the PR #887 and #785.

Thanks!

File: src/bin/vhost_user_fs.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ use futures::executor::{ThreadPool, ThreadPoolBuilder};
 use libc::EFD_NONBLOCK;
 use log::*;
 use std::num::Wrapping;
```
Comment (rbradford) [inline src/bin/vhost_user_fs.rs:17]: Commit message. s/congestion/contention/

File: vhost_user_net/src/lib.rs [hunk 2]
```diff
@@ -96,97 +96,64 @@ struct VhostUserNetThread {
     mem: Option<GuestMemoryMmap>,
```
Comment (rbradford) [inline vhost_user_net/src/lib.rs:95]: High level outline in the commit message about what you split?

File: vhost_user_net/src/lib.rs [hunk 3]
```diff
@@ -473,15 +451,19 @@ pub fn start_net_backend(backend_command: &str) {
     )
     .unwrap();
 
-    let vring_worker = net_daemon.get_vring_workers();
+    let mut vring_workers = net_daemon.get_vring_workers();
+
+    if vring_workers.len() != net_backend.read().unwrap().threads.len() {
+        println!("Number of vring workers must be identical to the number of backend threads");
```
Comment (rbradford) [inline vhost_user_net/src/lib.rs:454]: `error!` or `eprintln!`

File: vhost_user_net/src/lib.rs [hunk 4]
```diff
@@ -337,17 +337,17 @@ impl VhostUserBackend for VhostUserNetBackend {
         Ok(false)
     }
 
-    fn exit_event(&self) -> Option<(EventFd, Option<u16>)> {
-        let tap_end_index = (self.num_queues + self.num_queues / 2 - 1) as u16;
-        let kill_index = tap_end_index + 1;
+    fn exit_event(&self, thread_index: usize) -> Option<(EventFd, Option<u16>)> {
+        // The exit event is placed after the queues and the tap event, which
```
Comment (rbradford) [inline vhost_user_net/src/lib.rs:334]: s/which/which is/

**Comment by sboeuf** [comment]
@rbradford comments addressed, PTAL

**Comment by rbradford** [comment]
Still lgmt. Merge when green!

**Pull request merged**